src/json-observe.ts

```diff
@@ -0,0 +1,262 @@
+/*!
+ * https://github.com/Starcounter-Jack/JSON-Patch
+ * json-patch-duplex.js version: 1.1.2
+ * (c) 2013 Joachim Wester
+ * MIT license
+ */
+declare var Proxy: any;
+declare var Reflect: any;
+declare var module: any;
+
+function _isArray(obj) { return Array.isArray ? Array.isArray(obj) : function (obj) { return obj.push && typeof obj.length === 'number' } };
+//3x faster than cached /^\d+$/.test(str)
+function isInteger(str: string): boolean {
+  var i = 0;
+  var len = str.length;
+  var charCode;
+  while (i < len) {
+    charCode = str.charCodeAt(i);
+    if (charCode >= 48 && charCode <= 57) {
+      i++;
+      continue;
+    }
+    return false; 
+  }
+  return true;
+}
+function deepClone(obj: any) {
+  switch (typeof obj) {
+    case "object":
+      return JSON.parse(JSON.stringify(obj)); //Faster than ES5 clone - http://jsperf.com/deep-cloning-of-objects/5
+
+    case "undefined":
+      return null; //this is how JSON.stringify behaves for array items
+
+    default:
+      return obj; //no need to clone primitives
+  }
+}
+
+class JsonObserver {
+
+  private originalObject;
+  private exposedObject: Object;
+  private cachedProxy: Object;
+  private patches: Array<Object>;
+  private isRecording: boolean;
+  private defaultCallback: Function;
+  public userCallback: Function;
+  public disableCallback: boolean;
+
+  escapePathComponent(str) {
+    if (str.indexOf('/') === -1 && str.indexOf('~') === -1) return str;
+    return str.replace(/~/g, '~0').replace(/\//g, '~1');
+  }
+  generateProxyAtPath(obj, path) {
+
+    if (!obj) return obj;
+
+    path = path === '/' ? '' : path; //root
+    var instance = this;
+    let proxy = new Proxy(obj,
+      {
+        get: (target, propKey, receiver) => {
+          if (propKey.toString() === '_proxy') return true; //to distinguish proxies
+          return Reflect.get(target, propKey, receiver);
+        },
+        set: function (target, key, receiver) {
+
+          //https://github.com/Starcounter-Jack/JSON-Patch/issues/125
+          if(typeof receiver === 'function')
```

Comment:
As I stated there https://github.com/Starcounter-Jack/JSON-Patch/issues/125 I'd sacrifice non-compliant JSONs support for better performance. That's why, I would not bother about it here, or at least change it in separate PR in bot duplex and observer, to evaluate the cost.